### ISSUE TITLE
Developing OpenCart extensions using Docker.

### ISSUE THREAD

**Issue by deezznutz** [bitnami/bitnami-docker-opencart#98]
I'd like to develop an extension for OpenCart using Docker.  I've downloaded the bitnami-docker-opencart image and fired it up, but I need access to OpenCart's files and I need to be able to edit them.  I was thinking I'd make a volume that points an html folder on my machine to the /opt/bitnami/opencart/ folder in the container.  So I created my html file, copied the contents of /opt/bitnami/opencart/ to my html folder, and modifed the docker-compose.yml file to have this:
    volumes:
      - './html/:/bitnami/opencart'

I've deleted the old volume and rebuilt my containers.  Now, when I bring them up, I get this in the Open Cart logs:

Restoring persisted OpenCart installation
realpath: /bitnami/opencart/config.php: Too many levels of symbolic links

It looks like this is a broken link which makes sense since it came out of the container.

What am I doing wrong?   Is there some other approach to doing web development with Docker?

**Comment by marcosbc** [comment]
Hi @deezznutz, by default we are only persisting certain files of OpenCart. However, it is possible to persist the entire `/opt/bitnami/opencart` directory, i.e. by modifying your `docker-compose.yml` like this:

```diff
@@ -20,8 +20,9 @@
       - OPENCART_DATABASE_USER=bn_opencart
       - OPENCART_DATABASE_NAME=bitnami_opencart
       - ALLOW_EMPTY_PASSWORD=yes
+      - OPENCART_DATA_TO_PERSIST=/opt/bitnami/opencart
     volumes:
-      - 'opencart_data:/bitnami/opencart'
+      - './opencart_data:/bitnami/opencart'
       - 'opencart_storage_data:/bitnami/opencart_storage/'
     depends_on:
       - mariadb
```

This also requires that you create the `opencart_data` directory (in your case, `html`) with RW privileges for the non-root `1001` user:

```console
$ mkdir opencart_data
$ sudo chown 1001:root opencart_data
```

Once deployed, you should see all files there:

```console
$ ls opencart_data
admin  catalog  config-dist.php  config.php  image  index.php  licenses  license.txt  php.ini  system
```

Hope it works for you!

**Comment by deezznutz** [comment]
Thanks for the reply @marcosbc!

I had missed the OPENCART_DATA_TO_PERSIST line, and I deleted everything in my html folder so the container could persist its files to the folder.

And it works!!!

Woo Hoo!

Thank you.

**Comment by deezznutz** [comment]
> Hi @deezznutz, by default we are only persisting certain files of OpenCart. However, it is possible to persist the entire `/opt/bitnami/opencart` directory, i.e. by modifying your `docker-compose.yml` like this:
> 
> ```diff
> @@ -20,8 +20,9 @@
>        - OPENCART_DATABASE_USER=bn_opencart
>        - OPENCART_DATABASE_NAME=bitnami_opencart
>        - ALLOW_EMPTY_PASSWORD=yes
> +      - OPENCART_DATA_TO_PERSIST=/opt/bitnami/opencart
>      volumes:
> -      - 'opencart_data:/bitnami/opencart'
> +      - './opencart_data:/bitnami/opencart'
>        - 'opencart_storage_data:/bitnami/opencart_storage/'
>      depends_on:
>        - mariadb
> ```
> 
> This also requires that you create the `opencart_data` directory (in your case, `html`) with RW privileges for the non-root `1001` user:
> 
> ```
> $ mkdir opencart_data
> $ sudo chown 1001:root opencart_data
> ```
> 
> Once deployed, you should see all files there:
> 
> ```
> $ ls opencart_data
> admin  catalog  config-dist.php  config.php  image  index.php  licenses  license.txt  php.ini  system
> ```
> 
> Hope it works for you!

**Comment by deezznutz** [comment]
One more question....with this setup, can I run OpenCart in Docker, run VSCode on my machine, edit files, and have the changes show up in my browser?

**Comment by marcosbc** [comment]
That would work, sure. Just note that you may need to disable OpCache via `PHP_ENABLE_OPCACHE=no`, or some files may get cached.

**Comment by deezznutz** [comment]
OK thanks, and finally (I hope!)....XDebug.  Any pointers on getting that set up?

**Comment by marcosbc** [comment]
To enable XDebug, you would need to edit `/opt/bitnami/php/etc/php.ini` to your needs, either by [extending the container image](https://github.com/bitnami/bitnami-docker-opencart#extend-this-image) or by mounting the configuration file. These are the currently related configuration lines, which are commented:

```ini
;[XDebug]
;zend_extension = xdebug
;xdebug.mode = debug
;xdebug.client_host = 127.0.0.1
;xdebug.client_port = 9000
;xdebug.output_dir = /tmp
;xdebug.remote_handler = dbgp
```

Technically it should be enough to uncomment these lines:

```ini
[XDebug]
zend_extension = xdebug
```

Note: If you mount the configuration file, you would also have to disable OpCache manually (`opcache.enable = 0`) because it would make the file non-writable.

**Comment by deezznutz** [comment]
Getting closer!

I've created a dockerfile
[dockerfile.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6351301/dockerfile.txt)

and I've modified my docker-compose to use the new image:

[docker-compose.yml.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6351320/docker-compose.yml.txt)

But now my opencart container can't connect to the database.

**Comment by marcosbc** [comment]
@deezznutz Could you share more info about the error? If you enable `BITNAMI_DEBUG=true` it should show more verbose logs.

Could it be that the "html" folder is not empty, but the MariaDB volume is empty? If so, it would try to connect to an existing database because it sees that there is an already initialized install. If so, I would recommend to re-initialize everything.

**Comment by deezznutz** [comment]
I added the debug line and rebuilt my containers.  Here's the error:


Fatal error: Uncaught Exception: Error: Table 'bitnami_opencart.oc_translation' doesn't exist<br />Error No: 1146<br />SELECT * FROM oc_translation WHERE store_id = '0' AND language_id = '0' AND (route = 'engb' OR route = '') in /bitnami/opencart/system/library/db/mysqli.php:40 Stack trace: #0 /bitnami/opencart/system/library/db.php(45): DB\MySQLi->query('SELECT * FROM o...') #1 /bitnami/opencart/catalog/model/design/translation.php(6): DB->query('SELECT * FROM o...') #2 /bitnami/opencart/system/engine/loader.php(248): ModelDesignTranslation->getTranslations('engb') #3 /bitnami/opencart/system/engine/proxy.php(47): Loader->{closure}(Array, Array) #4 /bitnami/opencart/catalog/controller/event/translation.php(6): Proxy->__call('getTranslations', Array) #5 /bitnami/opencart/system/engine/action.php(79): ControllerEventTranslation->index('engb', '', Array) #6 /bitnami/opencart/system/engine/event.php(62): Action->execute(Object(Registry), Array) #7 /bitnami/opencart/system/engine/loader.php(210): Event->trigger('language/engb in /bitnami/opencart/system/library/db/mysqli.php on line 40

**Comment by marcosbc** [comment]
Like I mentioned in my earlier comment, this could be happening because the "html" directory or "opencart_storage" volumes are not empty, while the MariaDB volume is empty. Then the app container will think it was already installed, while the MariaDB container will think it needs to be re-installed.

To avoid that, you need to remove all volumes and re-initialize everything. After that it should work fine, just make sure not to remove any of the volumes or change the name of the parent directory that contains your `docker-compose.yml`.

**Comment by deezznutz** [comment]
Thanks for all your help Marcos.  I've got my opencart image running with my web files visible outside the container and with XDebug.   Now, I want to make the /bitnami/opencart_storage/ folder visible outside the container.

Would I just add these lines to my docker-compose file?
    environment:      
      - OPENCART_DATA_TO_PERSIST=/opt/bitnami/opencart:/opt/bitnami/opencart_storage
    volumes:
      - './opencart:/bitnami/opencart'
      - './opencart_storage:/bitnami/opencart_storage/'

**Comment by deezznutz** [comment]
Well, it should be obvious by now that this is my first Docker project.

What I REALLY want to do is have a container running OpenCart and be able to develop an extension for OpenCart in that container.  I've been using VSCode, and I just found they have an extension that lets you do just that.  So all this making files available outside the container may not be necessary.  But when I fire up VSCode and tell it to attach to my running OpenCart container, I get an error:

Command in container failed: mkdir -p /.vscode-server/bin/3c4e3df.....

It seems I need to have HOME set to some folder in /home/, but I don't know what.
But even this may not work.  I bash'ed into my container and I couldn't create a folder in the /home folder.

Any ideas?  Any at all?

**Comment by marcosbc** [comment]
Hi @deezznutz, no problem! We're here to help for issues like these.

As for VS Code support, unfortunately I'm not familiar with that extnesion. But from the error, it looks like it is going to try and start a server in the container, for which it would require to be able to create files in the `/.vscode-server` directory.

You could try and update your Dockerfile to create that folder with 777 permissions, that would allow any process to write on top of it.

(P.S.: If you need to also make `opencart_storage` accessible in the host, you would need to re-initialize the volumes again, or to [manually backup `/bitnami/opencart_storage`](https://github.com/bitnami/bitnami-docker-opencart#step-2-run-the-backup-command) that folder and extract to your computer)

**Comment by deezznutz** [comment]
Is it normal to fix one thing only to break two others?  :)

Trying to get XDebug working, but now my opencart container can't connect to my mariadb container.  I don't think I changed anything that would break this.

Here's my dockerfile, docker-compose, and "docker container inspect" results from the two containers:
[dockerfile.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6367414/dockerfile.txt)
[docker-compose.yml.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6367417/docker-compose.yml.txt)
[myopencart.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6367486/myopencart.txt)
[mymariadb.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6367487/mymariadb.txt)

**Comment by deezznutz** [comment]
This is so bizarre.  After powering my computer on this morning, I thought maybe I was doing something wrong in my dockerfile when I changed users to create the /.vscode-server folder.  So I messed around with that a while which required recreating my image which also (I think) required destroying and recreating my containers/volumes/etc (which I have a script for now).  So I did all that and I'm pretty sure what I was doing in the dockerfile is correct:  if I add a "RUN whoami" line just before I set the USER to 0, I get:
whoami: cannot find name for user ID 1001

But I now I'm back to where I was earlier yesterday: my containers can talk to each other.  Woo hoo!  
I'm pretty sure in the course of things yesterday, docker's networking got messed up because I also have the bitnami opencart on my machine and when I tried bringing that up yesterday, the two containers couldn't talk to each other either.

Now when I bring up my container for the first time, during the install of opencart I get these errors:

opencart 11:14:17.56 INFO  ==> Running install script
Warning: Use of undefined constant DIR_STORAGE - assumed 'DIR_STORAGE' (this will throw an Error in a future version of PHP) in /opt/bitnami/opencart/system/startup.php on line 70
PHP Warning:  Use of undefined constant DIR_STORAGE - assumed 'DIR_STORAGE' (this will throw an Error in a future version of PHP) in /opt/bitnami/opencart/system/startup.php on line 70
chmod: changing permissions of '/opt/bitnami/opencart/image/': Operation not permitted
chmod: changing permissions of '/opt/bitnami/opencart/image/no_image.png': Operation not permitted
chmod: changing permissions of '/opt/bitnami/opencart/image/catalog': Operation not permitted
chmod: changing permissions of '/opt/bitnami/opencart/image/catalog/opencart.ico': Operation not permitted
chmod: changing permissions of '/opt/bitnami/opencart/image/catalog/cart.png': Operation not permitted
chmod: changing permissions of '/opt/bitnami/opencart/image/catalog/profile-pic.png': Operation not permitted
chmod: changing permissions of '/opt/bitnami/opencart/image/catalog/opencart-logo.png': Operation not permitted

As a result of this, my opencart store has no images.

**Comment by deezznutz** [comment]
I'm pretty sure I got the networking issue figured out....If my mariadb container comes up but my opencart container fails, I can't just do "docker-compose up" again.  I have to take everything down first, then bring it back up so they can all be on the same network.  Now just those pesky images...

**Comment by marcosbc** [comment]
>I'm pretty sure I got the networking issue figured out....If my mariadb container comes up but my opencart container fails, I can't just do "docker-compose up" again. I have to take everything down first, then bring it back up so they can all be on the same network. Now just those pesky images...

That's weird. If you are using `docker-compose up` it should do that for you automatically, even if you are starting a specific container.

About the images issue, I think I know what is going on. Note that we are disabling stdout for the installation command, but it is shown when running with `BITNAMI_DEBUG=true`. That error appears in any Bitnami OpenCart container image deployment when running as a non-root user.

So, when you run the installation command, folders have not yet been persisted into `/bitnami/opencart` so they are owned by `root:root`. Because of this, any `chmod` operation run by a different user (i.e. `1001`) will fail. Note that, a bit later the folders are persisted into `/bitnami/opencart`, at which point the ownership is changed to `1001:root`. At this point, any chmod operation would work.

However, `images` really does not seem to require such ownership changes (after being initialized with Bitnami permissions), and `system/storage` is really the name for the `opencart_storage` volume which is placed at a different folder. It seems like it sets proper permissions to images so that the web server can write files to it, although it is being too lax (giving o+w permissions meaning any user could write to them).

So, in summary, the warning is expected in non-root deployments when running with `BITNAMI_DEBUG=true`. And, in fact, those warnings will cause any harm, since the container image already takes care of setting proper ownership/permissions.

**Comment by deezznutz** [comment]
OK, but when I run my containers I still don't see any images except for the shopping cart icon at the top-right of the main page.

I can look at the ...opencart/image/catalog/demo/banners folder on my computer and see the two images that are supposed to be on the banner, but they're not there.

If I bash into my opencart container, I can see those two images and that they have RW permissions for everyone.

Looking at the source for the page, all the "<img" tags have src="".  

And stepping thru the code, because I can now, I see that the code is failing in catalog/model/tool/image.php:
Line 4 has this:
		if (!is_file(DIR_IMAGE . $filename) || substr(str_replace('\\', '/', realpath(DIR_IMAGE . $filename)), 0, strlen(DIR_IMAGE)) != str_replace('\\', '/', DIR_IMAGE)) {
			return;
		}

substr(str_replace('\\', '/', realpath(DIR_IMAGE . $filename)), 0, strlen(DIR_IMAGE)) == "/bitnami/opencart/image/cata"
but
str_replace('\\', '/', DIR_IMAGE) == "/opt/bitnami/opencart/image/"

It looks like this is because DIR_IMAGE has /opt/ at the front and realpath only returns the /bitnami/,,,, part.

So.... I edited /config.php to make DIR_IMAGE = "/bitnami/opencart/image/" and images work now!
Woo!

Is this something I should do in my dockerfile?

**Comment by marcosbc** [comment]
I see... It looks like a limitation of the OpenCart application, as it does not allow running it under symlinks. It seems like the installation command, by default, is setting `DIR_APPLICATION`, `DIR_SYSTEM` and `DIR_IMAGE` to the symlink path, instead of the actual destination in the file system.

To change that, you would need to override those values in `config.php` and `admin/config.php` after the first initialization (since the files are created during installation). One way to do that is via [init scripts](https://github.com/bitnami/bitnami-docker-opencart/blob/master/3/debian-10/rootfs/post-init.sh): By mounting a .sh script to make the replacement after the first initialization at `/docker-entrypoint-init.d`.

**Comment by deezznutz** [comment]
OK, I've added two script files to my dockerfile.  All are attached.  

The resulting config.php files look good to me, but I get an error when starting OpenCart:
[Mon Apr 26 15:34:04.568556 2021] [php7:error] [pid 206] [client 192.168.208.1:48942] PHP Fatal error:  require_once(): Failed opening required '/bitnami/opencart/system/startup.php' (include_path='.:/opt/bitnami/php/lib/php') in /opt/bitnami/opencart/index.php on line 17

And it's right:  
/bitnami/opencart/system/startup.php doesn't exist, but
/opt/bitnami/opencart/system/startup.php does.

Where from here?


 
[dockerfile.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6378371/dockerfile.txt)
[update-admin-config.sh.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6378372/update-admin-config.sh.txt)
[update-config.sh.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6378373/update-config.sh.txt)

**Comment by marcosbc** [comment]
Unfortunately we do not have enough information to determine what may be failing in your side. If you could provide also the `docker-compose.yml` file you are using as well as the logs from the app initialization scripts, it would be helpful.

However, you did mention this:

>  /bitnami/opencart/system/startup.php doesn't exist, but
> /opt/bitnami/opencart/system/startup.php does.

Could it be that you removed `OPENCART_DATA_TO_PERSIST=/opt/bitnami/opencart` from your Docker environment configuration? If so, that would explain why `system` is not present in the persisted volume.

**Comment by deezznutz** [comment]
Yeah, I've been trying different OPENCART_DATA_TO_PERSIST and different volumes: settings.  Each gives a different kind of error.  I've attached my latest one.

[docker-compose.yml.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6386066/docker-compose.yml.txt)

**Comment by marcosbc** [comment]
From your docker-compose I can see this:

```yaml
    volumes:
      - './opencart:/opt/bitnami/opencart'
```

Note that you should never mount volumes to `/opt/bitnami`, but `/bitnami`, where persisted volumes are supposed to be. So it should be using something like this:

```yaml
    volumes:
      - './opencart:/bitnami/opencart'
```

**Comment by marcosbc** [comment]
Another issue I could see when testing your docker-compose is that `OPENCART_DATA_TO_PERSIST` is not being used properly. After setting it like this, it started working:

```yaml
      - "OPENCART_DATA_TO_PERSIST=/opt/bitnami/opencart/ /bitnami/opencart_storage/"
```

**Comment by deezznutz** [comment]
Ah.  I had tried a space in the OPENCART_DATA_TO_PERSIST setting, but I hadn't tried wrapping the whole thing in quotes.

I've made the changes you suggested, but now my opencart and opencart_storage folders are not being populated.  I did run my "destroy" script before bringing things back up.


[docker-compose.yml.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6393569/docker-compose.yml.txt)

**Comment by marcosbc** [comment]
@deezznutz Did you set proper permissions/ownership to the folders? If not it is not going to work. If you continue finding issues just set the following property inside the `opencart` block so containers have privileges to create them:

```yaml
user: root
```

And as we mentioned before, if you provide logs, it may give an additional hint on how to solve these issues, as they contain info about the context of the error and in some cases even the specific reason for the failure.

**Comment by deezznutz** [comment]
Hi Marcos,

Yes, I only set the opencart and opencart_storage folders to be read/write for everyone.  I was still the owner.  I've corrected that.

Here's how I build my image:
[buildmyimage.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6400756/buildmyimage.txt)
[dockerfile.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6400757/dockerfile.txt)

And here's the output of the buildmyimage command:
[build_output.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6400764/build_output.txt)

Here's my docker-compose file (with user: root added) and the script I use to run that:
[docker-compose.yml.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6400829/docker-compose.yml.txt)
[myopencartpersistup.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6400768/myopencartpersistup.txt)
[myopencartpersistup_output.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6400784/myopencartpersistup_output.txt)

And here are the permissions on my opencart and opencart_storage folders:
[folder-permissions.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6400804/folder-permissions.txt)

And permissions within the opencart folder:
[opencart-folder-permissions.txt](https://github.com/bitnami/bitnami-docker-opencart/files/6400835/opencart-folder-permissions.txt)

The opencart_storage folder is empty.

**Comment by deezznutz** [comment]
I've documented [my development environment](https://wzjz.us/index.php?/archives/42-Developing-OpenCart-Extensions-in-Docker.html) in case anyone wants to play along.

**Comment by marcosbc** [comment]
Hi @deezznutz, in your docker-compose I can see this:

```
      - "OPENCART_DATA_TO_PERSIST=/opt/bitnami/opencart/ /bitnami/opencart_storage/"
```

Note that `OPENCART_DATA_TO_PERSIST` only supports specifying paths to files inside `/opt/bitnami/opencart`. In your case, since `opencart_storage` will be persisted anyways (because you are mounting `/bitnami/opencart_storage` in the persisted volume) you only need to specify this:

```yaml
      - "OPENCART_DATA_TO_PERSIST=/opt/bitnami/opencart/"
```

I'm not sure how I got that specifically working though, but it should be enough to only specify the above path, given you also include this in docker-compose:

```yaml
    volumes:
      - './opencart:/bitnami/opencart'
      - './opencart_storage:/bitnami/opencart_storage/'
```

**Comment by deezznutz** [comment]
Thanks Marcos.  I was hoping to be able to just copy files from my computer into the opencart_storage folder outside the container, but I can just as easily copy them into the container.

Thanks for all your help.  I think I'm finally up and running with OpenCart extension development in Docker!

I definitely owe you a beer or six!

Walt

**Comment by marcosbc** [comment]
It's great that you got it working! Don't hesitate to contact us if you find any related issues in the future.

**Comment by symeonmanis** [comment]
Hello @marcosbc, I have the same issue with images not shown in my frontend. The opencart_data folder has 1001:root permissions.
Here is my dockerfile
[docker-compose.txt](https://github.com/bitnami/bitnami-docker-opencart/files/7815409/docker-compose.txt)

**Comment by carrodher** [comment]
Please, [note the Bitnami team is not actively monitoring GitHub Issues for this project](https://github.com/bitnami/bitnami-docker-opencart#community-supported-solution). This issue will stay open for 20 days to allow the community to contribute, after 15 days without activity the issue will be marked as stale being closed after 5 days.

If you would like to contribute to this container image by fixing a bug or adding a new feature feel free to create a PR, the Bitnami team will be happy to review it and provide feedback.
